### ISSUE TITLE
Memory configuration mismatch -> OOM

### ISSUE THREAD

**Issue by ctron** [EnMasseProject/enmasse#1451]
When scaling down consumers in my test I would expect the shareded queue brokers to buffer messages until consumer can catch up. That works as first, but quite soon brokers start to die due to an OOM situation.

Looking into the broker template I do see the following two parameters:

~~~yaml
- description: Memory limits for container
  name: BROKER_MEMORY_LIMIT
  value: 512Mi
- description: Global max size for all addresses in broker. Cannot be larger than
  a quarter of BROKER_MEMORY_LIMIT
  name: GLOBAL_MAX_SIZE
  value: 128Mb
~~~

Assuming the latter parameter targets the Java VM heap space this would be a problem because the startup script of the brokers already splits the `BROKER_MEMORY_LIMIT` in half in order to compensate for JVM overhead:

See the startup log:
~~~
-Xms256m -Xmx256m -XX:+UseParallelGC …
~~~

So I think the default settings for `BROKER_MEMORY_LIMIT` should be `1024Mi`instead.

**Comment by k-wall** [comment]
The Artemis brokers underlying the sharded queue are configured with an ```address-full-policy``` of [```FAIL```](https://activemq.apache.org/artemis/docs/2.5.0/paging.html), so I think the client should be seeing rejects once the shards are full rather than buffering.

Turning to the comment about the memory sizing, I agree, but wonder why [```-XX:+UseCGroupMemoryLimitForHeap```](https://blogs.oracle.com/java-platform-group/java-se-support-for-docker-cpu-and-memory-limits) isn't used.

**Comment by lulf** [comment]
@ctron @k-wall 

BROKER_MEMORY_LIMIT limits the memory given by the container for Artemis. The startup scripts for artemis will do the same as "-XX:+UseCGroupMemoryLimitForHeap " (I don't think it existed in Java 8 at some point), set the heap to half of the system memory limit, i.e. 256Mb.

GLOBAL_MAX_SIZE is not related to the heap, but is the max memory Artemis will use for its queues.

So: GLOBAL_MAX_SIZE < BROKER HEAP < BROKER_MEMORY_LIMIT

@ctron in your situation, I would expect the consumers queues to block any producers once their aggregate memory usage went beyond GLOBAL_MAX_SIZE. Since you still get an OOM, there must be something allocating memory off-heap, causing BROKER_HEAP + off-heap > BROKER_MEMORY_LIMIT.

**Comment by k-wall** [comment]
@ctron can you clarify "start to die due to an OOM situation".   Are you referring to a container OOM or are you seeing an JVM ```OutOfMemoryError```?  If the latter, what's the exception's description so we know whether the issue was heap or direct  (or even native threads).

@lulf doesn't ```FAIL`` give a dropped message behaviour rather than block?

**Comment by lulf** [comment]
@k-wall I think last i tried and understand what FAIL does, it responded with a detach if you tried to connect a sender to the queue, but it would still block if you were already attached. The behavior in Artemis isn't really expected iirc, but my memory is a bit vague.

**Comment by ctron** [comment]
> GLOBAL_MAX_SIZE is not related to the heap, but is the max memory Artemis will use for its queues.

True, but if that causes the broker to consume more memory that its JVM heap has available, it will trigger an OutOfMemory condition.

@k-wall IIRC the JVM killed itself due to OOM. But I will try to reproduce it.

**Comment by lulf** [comment]
@ctron Perhaps the memory required by other broker stuff than queues is > 384 MB, in which case you would benefit from increasing the heap memory potentially. If the memory is off-heap, it won't help.

That being said, I 'm thinking we should set GLOBAL_MAX_SIZE to -1, so that it is automatically tuned based on heap memory instead. I think exposing it with any other value is going to be error prone since its dependant on heap memory.

**Comment by ctron** [comment]
> @ctron Perhaps the memory required by other broker stuff than queues is > 384 MB, in which case you would benefit from increasing the heap memory potentially. If the memory is off-heap, it won't help.

Is > 128MB … as you can see the broker only get 256 MB assigned, so 128MB for the "global data" and 128 MB for the rest of the broker:

    -Xms256m -Xmx256m -XX:+UseParallelGC …

**Comment by lulf** [comment]
@ctron Ah sorry yes, you are right, 128MB for the rest of course. Clearly it would be useful if we could just leave global_max up to the broker :) (which I think it does if you set to -1). In any case, maybe we should update the defaults anyway, as I don't think this little memory for Artemis is well tested.

**Comment by k-wall** [comment]
https://github.com/EnMasseProject/enmasse/pull/1505 addressed one aspect of this issue.   Do we also want to make the memory default more generous too?  Double the JVM heap?

**Comment by lulf** [comment]
@k-wall We can try that, although last time I attempted to do that, I ran into an issue with kubernetes not having enough memory to schedule pods in travis CI. 

I think this can be fixed by changing the 'default' kubernetes configuration in CI, but I don't know the exact tunables.

**Comment by lulf** [comment]
I believe this is fixed in 0.26.3 and newer
